From 01d3ae030937b024bba63445cefd5050be4b2c8e Mon Sep 17 00:00:00 2001
From: Lars Uebernickel <lars.uebernickel@canonical.com>
Date: Thu, 1 Oct 2015 15:35:09 +0200
Subject: [PATCH] calendar: always emit "day-selected" once

"day-selected" was emitted twice when clicking on a day of the previous
or next month. This was because calendar_set_month{next,prev}() always
selected the first (or last) day of the newly focussed month. Callers
that wanted a different day to be selected had to do so after calling
one of those functions.

Fix this by passing the desired day into those functions directly.

https://bugzilla.gnome.org/show_bug.cgi?id=755941
---
 gtk/gtkcalendar.c | 72 +++++++++++++++++++++++++++++++++----------------------
 1 file changed, 43 insertions(+), 29 deletions(-)

diff --git a/gtk/gtkcalendar.c b/gtk/gtkcalendar.c
index b4ef543..2d338ef 100644
--- a/gtk/gtkcalendar.c
+++ b/gtk/gtkcalendar.c
@@ -876,11 +876,15 @@ calendar_queue_refresh (GtkCalendar *calendar)
 }
 
 static void
-calendar_set_month_next (GtkCalendar *calendar)
+calendar_set_month_next (GtkCalendar *calendar,
+                         gint         day)
 {
   gint month_len;
   GtkCalendarPrivate *priv = calendar->priv;
 
+  if (day < 0)
+    day = priv->selected_day;
+
   if (priv->display_flags & GTK_CALENDAR_NO_MONTH_CHANGE)
     return;
 
@@ -902,13 +906,13 @@ calendar_set_month_next (GtkCalendar *calendar)
 
   month_len = month_length[leap (priv->year)][priv->month + 1];
 
-  if (month_len < priv->selected_day)
+  if (month_len < day)
     {
       priv->selected_day = 0;
       gtk_calendar_select_day (calendar, month_len);
     }
   else
-    gtk_calendar_select_day (calendar, priv->selected_day);
+    gtk_calendar_select_day (calendar, day);
 
   calendar_queue_refresh (calendar);
 }
@@ -1298,11 +1302,15 @@ calendar_day_rectangle (GtkCalendar  *calendar,
 }
 
 static void
-calendar_set_month_prev (GtkCalendar *calendar)
+calendar_set_month_prev (GtkCalendar *calendar,
+                         gint         day)
 {
   GtkCalendarPrivate *priv = calendar->priv;
   gint month_len;
 
+  if (day < 0)
+    day = priv->selected_day;
+
   if (priv->display_flags & GTK_CALENDAR_NO_MONTH_CHANGE)
     return;
 
@@ -1325,16 +1333,16 @@ calendar_set_month_prev (GtkCalendar *calendar)
                  gtk_calendar_signals[MONTH_CHANGED_SIGNAL],
                  0);
 
-  if (month_len < priv->selected_day)
+  if (month_len < day)
     {
       priv->selected_day = 0;
       gtk_calendar_select_day (calendar, month_len);
     }
   else
     {
-      if (priv->selected_day < 0)
-        priv->selected_day = priv->selected_day + 1 + month_length[leap (priv->year)][priv->month + 1];
-      gtk_calendar_select_day (calendar, priv->selected_day);
+      if (day < 0)
+        day = day + 1 + month_length[leap (priv->year)][priv->month + 1];
+      gtk_calendar_select_day (calendar, day);
     }
 
   calendar_queue_refresh (calendar);
@@ -2811,10 +2819,10 @@ calendar_arrow_action (GtkCalendar *calendar,
       calendar_set_year_next (calendar);
       break;
     case ARROW_MONTH_LEFT:
-      calendar_set_month_prev (calendar);
+      calendar_set_month_prev (calendar, -1);
       break;
     case ARROW_MONTH_RIGHT:
-      calendar_set_month_next (calendar);
+      calendar_set_month_next (calendar, -1);
       break;
     default:;
       /* do nothing */
@@ -2913,21 +2921,27 @@ calendar_main_button_press (GtkCalendar    *calendar,
       day = priv->day[row][col];
 
       if (day_month == MONTH_PREV)
-        calendar_set_month_prev (calendar);
+        {
+          calendar_set_month_prev (calendar, day);
+        }
       else if (day_month == MONTH_NEXT)
-        calendar_set_month_next (calendar);
-
-      if (!gtk_widget_has_focus (widget))
-        gtk_widget_grab_focus (widget);
-
-      if (event->button == GDK_BUTTON_PRIMARY)
         {
-          priv->in_drag = 1;
-          priv->drag_start_x = x;
-          priv->drag_start_y = y;
+          calendar_set_month_next (calendar, day);
         }
+      else
+        {
+          if (!gtk_widget_has_focus (widget))
+            gtk_widget_grab_focus (widget);
+
+          if (event->button == GDK_BUTTON_PRIMARY)
+            {
+              priv->in_drag = 1;
+              priv->drag_start_x = x;
+              priv->drag_start_y = y;
+            }
 
-      calendar_select_and_focus_day (calendar, day);
+          calendar_select_and_focus_day (calendar, day);
+        }
     }
   else if (event->type == GDK_2BUTTON_PRESS)
     {
@@ -3098,13 +3112,13 @@ gtk_calendar_scroll (GtkWidget      *widget,
     {
       if (!gtk_widget_has_focus (widget))
         gtk_widget_grab_focus (widget);
-      calendar_set_month_prev (calendar);
+      calendar_set_month_prev (calendar, -1);
     }
   else if (event->direction == GDK_SCROLL_DOWN)
     {
       if (!gtk_widget_has_focus (widget))
         gtk_widget_grab_focus (widget);
-      calendar_set_month_next (calendar);
+      calendar_set_month_next (calendar, -1);
     }
   else
     return FALSE;
@@ -3179,7 +3193,7 @@ gtk_calendar_key_press (GtkWidget   *widget,
     case GDK_KEY_Left:
       return_val = TRUE;
       if (event->state & GDK_CONTROL_MASK)
-        calendar_set_month_prev (calendar);
+        calendar_set_month_prev (calendar, -1);
       else
         {
           move_focus (calendar, -1);
@@ -3192,7 +3206,7 @@ gtk_calendar_key_press (GtkWidget   *widget,
     case GDK_KEY_Right:
       return_val = TRUE;
       if (event->state & GDK_CONTROL_MASK)
-        calendar_set_month_next (calendar);
+        calendar_set_month_next (calendar, -1);
       else
         {
           move_focus (calendar, 1);
@@ -3246,11 +3260,11 @@ gtk_calendar_key_press (GtkWidget   *widget,
 
           day = priv->day[row][col];
           if (priv->day_month[row][col] == MONTH_PREV)
-            calendar_set_month_prev (calendar);
+            calendar_set_month_prev (calendar, day);
           else if (priv->day_month[row][col] == MONTH_NEXT)
-            calendar_set_month_next (calendar);
-
-          calendar_select_and_focus_day (calendar, day);
+            calendar_set_month_next (calendar, day);
+          else
+            calendar_select_and_focus_day (calendar, day);
         }
     }
 
-- 
2.5.0

